SwiftCompilerSources/Sources/Optimizer/Utilities/Cloner.swift

```diff
@@ -0,0 +1,186 @@
+//===--- Cloner.swift ------------------------------------------------------==//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import OptimizerBridging
+import SIL
+import SILBridging
+
+/// Clones the initializer value of a GlobalVariable.
+///
+/// Used to transitively clone "constant" instructions, including their operands,
+/// from or to the static initializer value of a GlobalVariable.
+///
+struct Cloner<Context: MutatingContext> {
+  private var bridged: BridgedCloner
+  let context: Context
+
+  enum Target {
+    case function(Function)
+    case global(GlobalVariable)
+  }
+  let target: Target
+
+  init(cloneToGlobal: GlobalVariable, _ context: Context) {
+    self.bridged = BridgedCloner(cloneToGlobal.bridged, context.bridgedPassContext)
+    self.context = context
+    self.target = .global(cloneToGlobal)
+  }
+
+  init(cloneBefore inst: Instruction, _ context: Context) {
+    self.bridged = BridgedCloner(inst.bridged, context.bridgedPassContext)
+    self.context = context
+    self.target = .function(inst.parentFunction)
+  }
+
+  init(cloneToEmptyFunction: Function, _ context: Context) where Context == FunctionPassContext {
+    self.bridged = BridgedCloner(cloneToEmptyFunction.bridged, context.bridgedPassContext)
+    self.context = context
+    self.target = .function(cloneToEmptyFunction)
+  }
+
+  mutating func deinitialize() {
+    bridged.destroy(context.bridgedPassContext)
+  }
+
+  var targetFunction: Function {
+    guard case .function(let function) = target else {
+      fatalError("expected cloning into a function")
+    }
+    return function
+  }
+
+  mutating func clone(instruction: Instruction) -> Instruction {
+    let cloned = bridged.clone(instruction.bridged).instruction
+    if case .function = target {
+      context.notifyInstructionChanged(cloned)
+      context.notifyInstructionsChanged()
+    }
+    return cloned
+  }
+
+  /// Transitively clones `value` including its defining instruction's operands.
+  mutating func cloneRecursively(value: Value) -> Value {
+    if isCloned(value: value) {
+      return getClonedValue(of: value)
+    }
+    
+    if let beginAccess = value as? BeginAccessInst {
```

Comment:
We should not hard code the handling of begin_access. Instead make it configurable with a closure